res/controllers/Denon-MC7000-scripts.js

```diff
@@ -815,26 +858,75 @@ MC7000.sortLibrary = function(channel, control, value) {
 /* LEDs for VuMeter */
 // VuMeters only for Channel 1-4 / Master is on Hardware
 MC7000.VuMeter = function(value, group) {
-    var deckNumber = script.deckFromGroup(group),
-        vuLevelOutValue = engine.getValue(group, "PeakIndicator") ? MC7000.VuMeterLEDPeakValue : Math.pow(value, 4) * (MC7000.VuMeterLEDPeakValue - 1);
-
-    midi.sendShortMsg(0xB0 + deckNumber - 1, 0x1F, vuLevelOutValue);
+    var deckNumber = script.deckFromGroup(group);
+    var currentLED = Math.pow(value, 3) * (MC7000.VuMeterLEDPeakValue - 1);
+    var vuLevelOutValue = engine.getValue(group, "PeakIndicator") ? MC7000.VuMeterLEDPeakValue : currentLED;
+    var vuThreshold = Math.floor(currentLED / 4) * 4;
+    // only send Midi signal when threshold is reached
+    if (currentLED < 4) {
+        midi.sendShortMsg(0xB0 + deckNumber - 1, 0x1F, 0x00);
+    } else { // switch off LEDs when currentLED < threashold
+        if (MC7000.prevVuLevel[deckNumber - 1] !== vuThreshold) {
```

Comment:
Can you explain what this threshold feature is for?

Reply:
The threshold is the distance between 2 midi signals sent to the VU meter. I just sketched this [picture for details](https://user-images.githubusercontent.com/58949409/112713274-a6591900-8ed4-11eb-8a9b-ea87a055c9f5.jpg). The next LED is only triggered every 12 values. So if the calculation is within this range, then no LED is changed but many midi signals are sent, very many signals at low latency times. I thought that it would be a good trade off to keep 3 signals sent in order to keep the resolution and don't shift too much.

Reply:
Thanks for the sketch it helped a lot. Why do you calculate the threshold on the vumeter value? wouldn't it just suffice to see whether the LED changes and only send an update on actual LED change? This would lead to the same visual quality but less code complexity and even less "traffic".

Reply:
I am open for code suggestions

Reply:
How about this snippet?

```suggestion
    var deckNumber = script.deckFromGroup(group);
    // scales value (0..1) to (0..117) while truncating to each LED
    var currentLED = Math.ceil(value * 9) * 13;
    var vuLevelOutValue = engine.getValue(group, "PeakIndicator") ? MC7000.VuMeterLEDPeakValue : currentLED;
    // only send Midi signal when LED value has changed
    if (MC7000.prevVuLevel[deckNumber - 1] != vuLevelOutValue) {
        midi.sendShortMsg(0xB0 + deckNumber - 1, 0x1F, vuLevelOutValue);
        MC7000.prevVuLevel[deckNumber - 1] = vuLevelOutValue;
    }
```
remove the for lines below

Reply:
unfortunately that doesn't work so well...
the `Math.ceil` makes the first LED always illuminated as the `MC7000.prevVuLevel[deckNumber - 1]` never reach `0`. 
Same on the other end. If there is not too much up and down inside the track and i turn on the gain then if i do quickly the red LED illuminates, If i am turning gain carefully up to the end (how i would do to level the track volume during a set) then the red LED never illuminates while the track is badly clipping. The GUI constantly in red, the controller max yellow, never reaches red. but it was worth a try. Thanks!

Reply:
Thats strange. It should reach 0 when the volume also goes to 0 and the PeakIndicator is still used, so if it clips in mixxx, it should display that on the controller too.

Reply:
Minor nitpick, think it would make sense to set `MC7000.VuMeterLEDPeakValue` to 0x7F, instead of 0x76, though that shouldn't make a difference functionally.

Reply:
@Swiftb0y - so now i have got your snippet principally working:

a) The upper red LED works only correctly when restarting Mixxx. Just updating the .js file gives this strange behavior. I did test several time with my and your code - The PeakIndicator issue happens with both after update the .js file but not restarting Mixxx. Restarting Mixxx then everything works as expected.

b) lower LED ... just changed your formula to show `currentLED = Math.floor(value * 9) * 13` so the last value is always 0 and not rounded to the next upper integer. This works great

Now I also remember why I introduced the threshold...
your formula gives a linear relationship between the signal strength and LED display. But this doesn't match to the Mixxx GUI nor to the Master Volume LEDs on the controller. The best fit to the GUI is with my formula `Math.pow(value, **4**) * 117` but I have got a few comments that some find that too nervous in the upper range and even one person wanted to have more information during a break in the lower area, so I changed that to `Math.pow(value, **3**) * 117` then the rise is not too drastic in the end of the scale but still matches optically quite well to the Master level that is handled by the controller. Now, when I insert the value³ formula again then there is no filter anymore and so a lot of values get sent to the controller. If you have got any other idea for the formula that gives a non-linear relationship is highly appreciated. Otherwise I introduce the threshold again. Hope this is made clear. One would think the matter behind the code is easy but in fact that was the tricky part for the VU meter ;)

Reply:
I just trying myself to find sth useful and visualized it - so I can share with you as well.... This is showing the LED MIDI Signal over the VuMeter value.
![grafik](https://user-images.githubusercontent.com/58949409/112761559-229a4c00-8ffc-11eb-8e02-389385daa4d6.png)

Reply:
Alternatively I could set like 9 times an `if` `else if` condition to check for `var currentLED` is inside the steps and only send signal if it goes outside each step. But I think that is overkill

Reply:
I am thinking too much difficult. This formula works easily...

`var currentLED = Math.floor(Math.pow(value,2.5) * 9) * 13;`

which gives the result shown on the red graph. So now i have the behavior that I wanted combined with less signals sent...

![grafik](https://user-images.githubusercontent.com/58949409/112762928-9d666580-9002-11eb-979e-3a056d52b446.png)

Reply:
> your formula gives a linear relationship between the signal strength and LED display. But this doesn't match to the Mixxx GUI nor to the Master Volume LEDs on the controller.

Ok thats strange, I was assuming this was the case. You can still compose both of our solutions together:
```javascript
var numSteps = 10;
var pow = 3;
value = Math.floor(Math.pow(x, pow)*numSteps)/numSteps) * 130;
var vuLevelOutValue = engine.getValue(group, "PeakIndicator") ? MC7000.VuMeterLEDPeakValue : value ;
```
When using `Math.floor` you probably want 10 steps, otherwise the last yellow LED will only be lit when the signal is at 1.0.
The above code will produce the yellow graph with the floor truncation, while the green graph is your original function:
![canvas](https://user-images.githubusercontent.com/12380386/112762780-dce08200-9001-11eb-90d6-3412f6f83881.png)

Reply:
As an optimization step, you can inline the calculation so it only gets executed when its actually needed.
```javascript
var vuLevelOutValue = engine.getValue(group, "PeakIndicator") ? 
    MC7000.VuMeterLEDPeakValue : 
    Math.floor(Math.pow(value, pow)*numSteps)/numSteps) * 130;
```

Reply:
In the case of 10 steps, it could even be simplified to `Math.floor(Math.pow(value, pow)*numSteps) * 13`, omitting an unnecessary division.

Reply:
> I am thinking too much difficult. This formula works easily...
> `var currentLED = Math.floor(Math.pow(value,2.5) * 9) * 13;`

Right, I somehow must've missed that comment. The only difference now is that I fear `* 9` will still produce subtly inaccurate values. Use `* 10` instead.

Reply:
and raising to the power of 2,3 instead of 3 also doesn't make much of a difference, since the controller will put it into each "led bucket" anyways, regardless of the exact value in the bucket.

Reply:
Hey @Swiftb0y - the factor `*9` is the right one. each segment calculates the last MIDI signal so the numbers are 13,26,39,52,65,78,91,104,117 always trigger the correct LED. IN regards to the last one only flashes at peak 117 is also very good. In this way it aligns to the GUI - you can have the yellow area dipping the top or even staying at the top without clipping. This behavior is alright. Another point about the exponent, it makes a big difference visually if you use 1,2,3,4 or 2,5. As stated before, the exp. 4 was matching the GUI quite well but gets very sensitive in the upper area. exp. 3 is matching okay but 2,5 is even better. With exp 2 the LEDs are not even close to the Controller Master nor the GUI Meters. No exponent is perfect though but 2,5 is quite nicely matching and gives good information to align the track level. I could make that adjustable in a user setting variable at the top but I think it shouldn't. ...

Reply:
Ok, if you like the behavior of `Math.floor(Math.pow(value,2.5) * 9) * 13;` that should also work.